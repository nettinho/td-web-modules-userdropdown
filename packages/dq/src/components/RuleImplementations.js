import _ from "lodash/fp";
import React from "react";
import { compose } from "redux";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { Table, Label, Icon } from "semantic-ui-react";
import { injectIntl, FormattedMessage } from "react-intl";

export const RuleImplementationRow = ({
  id,
  implementation_key,
  description,
  system,
  rule,
  system_params,
  rule_results,
  minimum,
  goal,
  intl: { formatMessage }
}) => {
  const result = _.isEmpty(rule_results) ? null : rule_results[0].result;
  const date = _.isEmpty(rule_results) ? null : rule_results[0].date;
  const resultTitle = !_.isNil(result)
    ? formatMessage(
        { id: "quality.rule.quality.description" },
        { result: result, date: date }
      )
    : formatMessage({ id: "quality.rule.quality.no.data" });
  return (
    <Table.Row key={id}>
      <Table.Cell content={implementation_key} />
      <Table.Cell content={_.truncate({ length: 60 })(description)} />
      <Table.Cell content={system} />
      <Table.Cell
        content={formatMessage({ id: `rule.type.${rule.rule_type.name}` })}
      />
      <Table.Cell
        content={Object.keys(system_params).map(function(key) {
          const value = system_params[key];
          return (
            <Label key={key}>
              {key}:{value}
            </Label>
          );
        })}
      />
      <Table.Cell>
        {_.isNil(result) ? (
          <Icon name="circle" title={resultTitle} color="grey" />
        ) : (
          <Icon name="circle" size="mini" color="grey" />
        )}
        {!_.isNil(result) && result < minimum ? (
          <Icon title={resultTitle} name="circle" color="red" />
        ) : (
          <Icon name="circle" size="mini" color="red" />
        )}
        {!_.isNil(result) && minimum <= result && result < goal ? (
          <Icon title={resultTitle} name="circle" color="yellow" />
        ) : (
          <Icon name="circle" size="mini" color="yellow" />
        )}
        {!_.isNil(result) && result >= goal ? (
          <Icon name="circle" title={resultTitle} color="green" />
        ) : (
          <Icon name="circle" size="mini" color="green" />
        )}
      </Table.Cell>
    </Table.Row>
  );
};

RuleImplementationRow.propTypes = {
  id: PropTypes.number,
  implementation_key: PropTypes.string,
  description: PropTypes.string,
  system: PropTypes.string,
  rule: PropTypes.object,
  system_params: PropTypes.object,
  intl: PropTypes.object,
  rule_results: PropTypes.array,
  minimum: PropTypes.number,
  goal: PropTypes.number
};

export const RuleImplementations = ({ rule, ruleImplementations, ...rest }) => {
  const { minimum, goal } = rule;
  return (
    <React.Fragment>
      {!_.isEmpty(rule) &&
        !_.isEmpty(ruleImplementations) && (
          <Table celled striped compact selectable>
            <Table.Header>
              <Table.Row>
                <Table.HeaderCell
                  content={
                    <FormattedMessage id="ruleImplementation.props.name" />
                  }
                />
                <Table.HeaderCell
                  content={<FormattedMessage id="quality.rule.description" />}
                />
                <Table.HeaderCell
                  content={<FormattedMessage id="quality.rule.system" />}
                />
                <Table.HeaderCell
                  content={<FormattedMessage id="quality.rule.type" />}
                />
                <Table.HeaderCell
                  content={<FormattedMessage id="quality.rule.system_params" />}
                />
                <Table.HeaderCell
                  content={<FormattedMessage id="quality.rule.quality" />}
                />
              </Table.Row>
            </Table.Header>
            <Table.Body>
              {ruleImplementations.map((s, i) => (
                <RuleImplementationRow
                  key={i}
                  {...{ ...s, minimum, goal }}
                  {...rest}
                />
              ))}
            </Table.Body>
          </Table>
        )}
    </React.Fragment>
  );
};

RuleImplementations.propTypes = {
  ruleImplementations: PropTypes.array,
  rule: PropTypes.object
};

const mapStateToProps = ({ rule, ruleImplementations }) => ({
  rule,
  ruleImplementations
});

export default compose(
  injectIntl,
  connect(mapStateToProps)
)(RuleImplementations);
