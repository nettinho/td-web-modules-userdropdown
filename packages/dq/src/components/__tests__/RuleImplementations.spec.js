import React from "react";
import { shallow } from "enzyme";
import { RuleImplementations } from "../RuleImplementations";

describe("<RuleImplementations />", () => {
  it("matches the latest snapshot", () => {
    const rule = {
      id: 123,
      business_concept_id: "2D2B3",
      name: "control1",
      description: "desc1",
      goal: 10,
      minimum: 1,
      principle: { name: "principle1" },
      type: "type1",
      type_params: {},
      status: "status1"
    };
    const ruleImplementations = [
      {
        id: 9,
        name: "NameRule1",
        description: "RuleDesc1",
        system: "SYS",
        type: "integer_values_range",
        system_params: { table: "Tabla1", column: "Columna1" }
      },
      {
        id: 10,
        name: "NameRule2",
        description: "RuleDesc2",
        system: "SYS",
        type: "integer_values_range",
        system_params: { table: "Tabla2", column: "Columna2" }
      }
    ];
    const props = {
      ruleImplementations,
      rule
    };

    const wrapper = shallow(<RuleImplementations {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
