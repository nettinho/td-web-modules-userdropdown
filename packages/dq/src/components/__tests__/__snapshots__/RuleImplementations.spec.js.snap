// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RuleImplementations /> matches the latest snapshot 1`] = `
<React.Fragment>
  <Table
    as="table"
    celled={true}
    compact={true}
    selectable={true}
    striped={true}
  >
    <TableHeader
      as="thead"
    >
      <TableRow
        as="tr"
        cellAs="td"
      >
        <TableHeaderCell
          as="th"
          content={
            <FormattedMessage
              id="ruleImplementation.props.name"
              values={Object {}}
            />
          }
        />
        <TableHeaderCell
          as="th"
          content={
            <FormattedMessage
              id="quality.rule.description"
              values={Object {}}
            />
          }
        />
        <TableHeaderCell
          as="th"
          content={
            <FormattedMessage
              id="quality.rule.system"
              values={Object {}}
            />
          }
        />
        <TableHeaderCell
          as="th"
          content={
            <FormattedMessage
              id="quality.rule.type"
              values={Object {}}
            />
          }
        />
        <TableHeaderCell
          as="th"
          content={
            <FormattedMessage
              id="quality.rule.system_params"
              values={Object {}}
            />
          }
        />
        <TableHeaderCell
          as="th"
          content={
            <FormattedMessage
              id="quality.rule.quality"
              values={Object {}}
            />
          }
        />
      </TableRow>
    </TableHeader>
    <TableBody
      as="tbody"
    >
      <RuleImplementationRow
        description="RuleDesc1"
        goal={10}
        id={9}
        key="0"
        minimum={1}
        name="NameRule1"
        system="SYS"
        system_params={
          Object {
            "column": "Columna1",
            "table": "Tabla1",
          }
        }
        type="integer_values_range"
      />
      <RuleImplementationRow
        description="RuleDesc2"
        goal={10}
        id={10}
        key="1"
        minimum={1}
        name="NameRule2"
        system="SYS"
        system_params={
          Object {
            "column": "Columna2",
            "table": "Tabla2",
          }
        }
        type="integer_values_range"
      />
    </TableBody>
  </Table>
</React.Fragment>
`;
